Added softban unban feature

MikeDX opened the pull request:
Short Description: 

Upon detection of a softban, bot will stop wasting balls and berries on pokemon and travel to a pokestop to try to unban itself.

Fixes:

Softban from travelling too far / quickly
Some small grammar changes (seem -> seems)

douglascamata commented:
@MikeDX what about extracting softban unban feature into a worker that runs every tick, checks `self.bot.softbanned` and run if it is true? Let's make things more modular!

DayBr3ak commented on pokemongo_bot/__init__.py:
```diff
@@ -26,6 +26,8 @@
 
 class PokemonGoBot(object):
 
+    softbanned = False
```
place this in the __init__ method ?

DayBr3ak commented on pokemongo_bot/cell_workers/seen_fort_worker.py:
```diff
@@ -108,14 +108,19 @@ def work(self):
+                            player_latitude=f2i(self.position[0]),
+                            player_longitude=f2i(self.position[1]))
+                        response_dict = self.api.call()
+
```
add a sleep ? seems much more human and less botty :)

MikeDX commented:
@douglascamata So, to do this:

if bot banned:
walk to pokestop
spin spin spin
set banned = false?

resume normal operation

douglascamata commented:
SoftBanFixWorker:
```
if bot banned:
    walk to pokestop (using MoveToFortWorker)
    spin once (using SeenFortWorker)
    if still banned:
        return WorkerState.RUNNING (so tick is immediately finished and next one starts)
```

Add this detail to SeenFortWorker
```
if got loot:
  set banned = false
```

And this SoftBanFixWorker should be added as first element of the worker list, to run first.

MikeDX commented:
Ok, makes sense. should the unban worker be in the root, or as a cell worker? (guessing root)

MikeDX commented:
@douglascamata How would you suggest we keep the nearest fort between ticks, since currently it is doing its job but moving between forts to try to fix each tick which isn't what we want.

MikeDX commented:
New worker created, removed lots of output when iterating through the spins. Works well, although takes about a minute to complete. 

Also checks for lures after the ban is lifted

MikeDX commented:
Resolved conflicts

wilsonyqm commented:
Can't wait to see this PR merged

MatthewARoy commented:
Ah excellent, I really hope this works as expected. It will be a very nice feature to auto-unban.
Thanks guys!

douglascamata commented:
@MikeDX actually we want it to recalculate the nearest fort at every tick. It should work without `self.bot.nearest_fort`.

MikeDX commented:
Using nearest every tick moves to the next one instead of spinning the same stop.

MikeDX commented:
rebased to dev + squashed again.

douglascamata commented:
@MikeDX but `SpinNearestFort` worker won't execute at all while you return `WorkerResult.RUNNING`, because you are the first worker

MikeDX commented:
I'm not calling that method or using SpinNearestFort anymore - it doesnt need it. The first time the bot runs, banned = false

then it goes through the list, eventually calling SpinNearestFortWorker which *does* set the nearest_fort value. Even if CatchVisiblePokemonWorker sets the ban to true before SpinNearest runs, SpinNearestFortWorker will still execute and find a stop regardless.

douglascamata commented:
@MikeDX hum, I see now. I think of another option other than the attribute `self.nearest_fort`: do your own fort search.

As simple as seen in `SpinNearestFortWorker`:

```
def nearest_pokestop(self):
    if 'forts' in self.bot.cell:
        forts = [fort
            for fort in self.bot.cell['forts']
             if 'latitude' in fort and 'type' in fort]
        forts.sort(key=lambda x: distance(self.position[
                       0], self.position[1], x['latitude'], x['longitude']))
        return forts[0]
    return None
```

douglascamata commented:
#1578 makes it even better! Fort search is not in `self.bot` and it doesn't include any filter. Perfect for you!

MikeDX commented:
@douglascamata Would have been handy to have that answer, when I asked the question 7 hours ago...

douglascamata commented:
#1578 `self.bot.get_forts` include ALL nearby forts, regardless of any filter. That's all you need.

TheSavior commented:
That function should probably be pulled out as well, we use it in a bunch of places. 

Side note, @douglascamata I think we need a new place for those types of functions that aren't the bot class.

douglascamata commented:
@TheSavior yes, we need refactor like #1253. But who wants to test/merge? :P

MikeDX commented:
@douglascamata I have updated the bot to get its own fort position, but since that function is not yet merged, I have replaced it with my own for now.

douglascamata commented:
@MikeDX ok... I do a quick update of your code in dev after we merge #1578

douglascamata commented:
👍

douglascamata commented:
@TheSavior give a thumbs up here... I update the SoftBanWorker to use your utility functions after we merge our PR.

TheSavior commented on pokemongo_bot/__init__.py:
```diff
@@ -77,6 +78,7 @@ def tick(self):
         self.check_session(self.position[0:2])
 
         workers = [
+            SoftBanWorker,
```
This should probably be down the list, right before catching pokemon. Since the bot can still transfer and evolve while banned.

douglascamata replied:
@TheSavior but which pokemon would you transfer or evolve if you cannot catch?

TheSavior replied:
That's a good point, at least with the current behavior of the workers we have.

TheSavior commented:
👍

TheSavior commented on pokemongo_bot/cell_workers/catch_visible_pokemon_worker.py:
```diff
@@ -15,7 +15,7 @@ def __init__(self, bot):
 
     def work(self):
-        if not self.config.catch_pokemon:
+        if not self.config.catch_pokemon or self.bot.softbanned:
```
we shouldn't need to change this

douglascamata replied:
why not?

douglascamata replied:
it will waste pokeballs if we don't do this

TheSavior replied:
It will never get to this worker since the softbanned worker returns RUNNING

MikeDX replied:
If there are more than one pokemon within range, yes it will

TheSavior commented:
:-1:

TheSavior commented on pokemongo_bot/cell_workers/move_to_fort_worker.py:
```diff
@@ -23,7 +23,8 @@ def work(self):
 
         if dist > 10:
-            logger.log('Moving towards fort {}, {} left'.format(fortID, format_dist(dist, unit)))
+            if self.bot.softbanned == False:
```
We shouldn't be changing this either.

douglascamata replied:
this one I agree

MikeDX replied:
This is about the only merge comment I agree with!

TheSavior commented on pokemongo_bot/cell_workers/soft_ban_worker.py:
```diff
@@ -0,0 +1,31 @@
+           SeenFortWorker(nearest_fort, self.bot).work()
+
+           if self.bot.softbanned == True:
+               return WorkerResult.RUNNING
```
This will keep anything after it from running in the tick. This means that NO other worker needs to know about whether we are soft banned. They should not have any check in it at all.

MikeDX replied:
No, because if a catch worker sets banned to true (if a berry fails) then spinworker needs to know.

douglascamata replied:
catch worker will never set banned to true, because we only know we are banned when we spin pokestop

MikeDX replied:
If this is the case why does the berry fail and show a softban message?

TheSavior commented on pokemongo_bot/cell_workers/soft_ban_worker.py:
```diff
@@ -0,0 +1,31 @@
+    def work(self):
+       if self.bot.softbanned == True:
+           nearest_fort = self.nearest_pokestop()
+           MoveToFortWorker(nearest_fort, self.bot).work()
```
You don't need this. The only way to get into the softbanned mode is to have spun a fort and not gotten anything. That means you are close enough to spin it again.

MikeDX replied:
Not true. You can be softbanned by throwing a berry and failing, or even if a pokemon ran away at 100% catch rate.

TheSavior commented on pokemongo_bot/cell_workers/soft_ban_worker.py:
```diff
@@ -0,0 +1,31 @@
+           if self.bot.softbanned == True:
+               return WorkerResult.RUNNING
+ 
+    def nearest_pokestop(self):
```
This doesn't seem right...

MikeDX replied:
I've no idea what part you are even referring to. This is all based on suggestions from @douglascamata after the initial (small) commit was rejected for not being modular.

TheSavior commented on pokemongo_bot/cell_workers/soft_ban_worker.py:
```diff
@@ -0,0 +1,31 @@
+
+from pokemongo_bot.cell_workers.utils import distance
+from pokemongo_bot.worker_result import WorkerResult
+from pokemongo_bot.cell_workers import SeenFortWorker
```
You can combine these imports

douglascamata replied:
no no, he can't... all these 3 imports are importing from different inner-modules.

MikeDX replied:
You can, but it's pretty horrible coding practice and certainly not what PEP recommends.

TheSavior replied:
Line 5 and 6 both import from `pokemongo_bot.cell_workers`

MikeDX replied:
line 6 isnt even in your comment

MikeDX replied:
line 6 isnt even in your comment

TheSavior commented on pokemongo_bot/cell_workers/seen_fort_worker.py:
```diff
@@ -74,15 +61,18 @@ def work(self):
+                    logger.log("Loot: ", 'green')
                     logger.log(str(experience_awarded) + " xp",
                                'green')
+                    if self.bot.softbanned == True:
```
The only thing that should set us to *not* be softbanned anymore should be the softbanned worker.

MikeDX replied:
How would that even work?

douglascamata replied:
Using `return WorkerResult.SUCCESS`

TheSavior commented on pokemongo_bot/cell_workers/pokemon_catch_worker.py:
```diff
@@ -122,9 +122,12 @@ def work(self):
+                                        logger.log('Failed to use berry. Seems like you are softbanned.', 'red')
+                                        logger.log('Heading to nearest Pokestop to attempt unban.', 'red')
+                                        self.bot.softbanned = True
+                                        break
```
Why isn't this a return? If we are softbanned, what else is there to do in the catch worker?

MikeDX commented:
@TheSavior @douglascamata Can I suggest one of you take what I've done and do all the testing and re-implementation of what I have done since you don't agree with any of this pull request.

MatthewARoy commented:
Thanks for working on it @MikeDX
Have we done any testing yet? I've seen it on a few websites but wanted to make sure with more thorough testing. If you get a branch in working order I'll check it out, make a new account and softban it a few times to test

MikeDX commented:
I did a lot of testing yesterday and spent many hours on it getting it to work as it should and bent it to fit "the rules"

Since closing this pull request, dev has changed multiple times and a lot has changed meaning most if not all would have to be re-done. I made a lot of changes to rebase with dev, and it still isn't working properly. One worker returns "11" or "0" - whatever that means, certainly not a SUCCESS or RUNNING value. 

As with my previous pull request, I'm not really willing to waste any more time on it for it to be glossed over and nitpicked. See above comments on the code - almost every single one was based upon a misunderstanding of what the problem was and how to fix it.

I have participated in multiple open source projects in the past and I have never seen such a crazy set of rules for merging code before - it is totally hit and miss here. I see garbage merged into the code that breaks the whole system, and even with syntax errors with tabulation, yet when I submit working, tested code, it is pulled apart for not being done in a certain way - and then nitpicked again by someone who did not even understand the problem in the first place. Frustrating and absolutely crazy.

TheSavior commented:
@MikeDX, I hear you and I definitely agree, I did not understand the problem and your solution as a whole while I was reviewing it. Our goal on this project is to be as inclusive as possible and to help contributors feel as good as possible contributing. We failed on this one, and that is entirely my fault.

We definitely want to get this feature in, it is valuable to many people using this bot. I understand the lack of interest you have iterating on this more without a clear path to success. Recognizing that, we will probably need to spend our time doing it. Because you deleted your branch, we will probably have to pull out the functionality from the diff itself. 

We are overwhelmed with pull requests and tend to do a first pass on reviews just pattern matching on the common issues we've seen. Typically that includes things like: implementing a feature that requires changes to lots of workers, workers calling other workers, PRs that say they do one thing and then actually change a bunch of other things as well. Unfortunately, I pattern matched this PR to the first case, which was not fair to this PR and the work you've done. When spending more time looking at it and making sure I understand the things actually being done, it makes more sense. 

Again, this is my mistake and thus I need to remedy that by getting this functionality added. I hope this experience won't completely put you off from this project, we always want to value quality contributions and quality contributors. Thank you for contributing this PR.

TheSavior commented:
@MikeDX Here is a follow up PR. Please take a look, your code influenced the implementation. https://github.com/PokemonGoF/PokemonGo-Bot/pull/1724

MikeDX commented:
@TheSavior I appreciate what you are saying, but pull request 1724 is shockingly bad and @douglascamata didn't even follow his own advice that he told me to do after my original commit!

I really hope you get your act(s) together on this project. It showed so much promise but arrogance and a certain level of ignorance has definitely put me off almost entirely now which for me is a real shame as there is absolutely no motivation to work on a feature or bug fix since submissions are so poorly vetted.

MikeDX commented:
See issue #2049. 

As I said before, I have contributed to many projects both open source and within professional environments and this is easily the worst managed one I have ever come across.

DayBr3ak commented:
"I have contributed so many time, I'm the bestttttt" Who is arrogant now ?
